---
title: "Within species analysis"
subtitle: ""
author: "lncRNA group"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: true
    toc_depth: 4
    toc_float:
      toc_collapsed: true
    number_sections: false
    theme: lumen
    df_print: paged
    code_folding: show
---
      
# Set the Project specific parameters
```{r setup, include=FALSE}
library("knitr")
library("GenomicTools")
library("Luke")
library("fastqcr")
library("factoextra")
library("Rtsne")
library("ggplot2")
options(scipen=999)
knitr::opts_chunk$set(echo = FALSE,
                      cache = TRUE,
                      cache.lazy = FALSE,
                      dev = c('png', 'pdf'),
                      fig.align = 'center', fig.height = 5, fig.width = 8.5)

bosFolder <- "/media/ejo138/Fantec6/Projects/FAANG-lncRNA/bos_taurus"
capraFolder <- "/media/ejo138/Fantec6/Projects/FAANG-lncRNA/"
equusFolder <- "/media/ejo138/Fantec6/Projects/FAANG-lncRNA/equus_caballus"
gallusFolder <- "/media/ejo138/Fantec6/Projects/FAANG-lncRNA/gallus_gallus"
ovisFolder <- "/media/ejo138/Fantec4/Projects/FAANG-lncRNA/ovis_aries"
susFolder <- "/media/ejo138/Fantec6/Projects/FAANG-lncRNA/sus_scrofa"
mainFolder <- "/media/ejo138/Fantec6/Projects/FAANG-lncRNA/"
projFolder <- "/home/ejo138/gitPipelines/Pipeline-FAANG-lncRNA/"


bosSamples <- as.vector(as.matrix(read.table(file.path(bosFolder, "samples"), stringsAsFactors = FALSE)))
#capraSamples <- as.vector(as.matrix(read.table(file.path(bosFolder, "samples"), stringsAsFactors = FALSE)))
equusSamples <- as.vector(as.matrix(read.table(file.path(equusFolder, "samples"), stringsAsFactors = FALSE)))
gallusSamples <- as.vector(as.matrix(read.table(file.path(gallusFolder, "samples"), stringsAsFactors = FALSE)))
ovisSamples <- as.vector(as.matrix(read.table(file.path(bosFolder, "samples"), stringsAsFactors = FALSE)))
susSamples <- as.vector(as.matrix(read.table(file.path(susFolder, "samples"), stringsAsFactors = FALSE)))

species <- c("Bos T.", "Capra H.", "Equus C.", "Gallus G.", "Ovis A.", "Sus S.")
samples <- read.table(file.path(projFolder,"2019.04.02-SamplesList.tsv"), sep="\t", stringsAsFactors=FALSE, header=TRUE)
samples[,2] <- tolower(samples[,2])
samples[,3] <- tolower(samples[,3])
```

Import the annotations

```{r}
bosLNC <- importGTF(file.path(bosFolder,"FEELnc", "codpot", "feelnc_bos_taurus.codpot.lncRNA.gtf"), level="exon")
bosLNCIDs <- unique(bosLNC$gene_id)
bosAnnot <- importGTF(file.path(mainFolder,"annotations","bos_taurus","ARS-UCD1.2.95", "ARS-UCD1.2.95.gtf"), level="gene")
bosMSTR2Annot <- read.table(file.path(mainFolder, "mstr2ens", "bos_taurus_MSTRG_ensembl.txt"), stringsAsFactors=FALSE)

#capraLNC <- importGTF(file.path(capraFolder,"FEELnc", "codpot", "feelnc_capra_hircus.codpot.lncRNA.gtf"), level="exon")
#capraLNCIDs <- unique(capraLNC$gene_id)

equusLNC <- importGTF(file.path(equusFolder,"FEELnc", "codpot", "feelnc_equus_caballus.codpot.lncRNA.gtf"), level="exon")
equusLNCIDs <- unique(equusLNC$gene_id)

gallusLNC <- importGTF(file.path(gallusFolder,"FEELnc", "codpot", "feelnc_gallus_gallus.codpot.lncRNA.gtf"), level="exon")
gallusLNCIDs <- unique(gallusLNC$gene_id)

ovisLNC <- importGTF(file.path(ovisFolder,"FEELnc", "codpot", "feelnc_ovis_aries.codpot.lncRNA.gtf"), level="exon")
ovisLNCIDs <- unique(ovisLNC$gene_id)

susLNC <- importGTF(file.path(susFolder,"FEELnc", "codpot", "feelnc_sus_scrofa.codpot.lncRNA.gtf"), level="exon")
susLNCIDs <- unique(susLNC$gene_id)
```

# Basic statistics

## Quality checks with FastQC

First, we import the quality checks from FastQC

```{r importFastQCReports, echo=FALSE, message=FALSE, warning=FALSE}
# Import the raw quality information
importRawReads <- function(dir){
  qcRawFiles <- list.files(file.path(dir,"FASTQC"), recursive=TRUE, pattern="*.zip")
  rawReads <- 0
  for(i in 1:length(qcRawFiles)){
    tmp <- qc_read(file.path(dir,"FASTQC",qcRawFiles[i]))
    rawReads[i] <- as.numeric(tmp$basic_statistics$Value[4]) 
  }
  names(rawReads) <- sapply(strsplit(sapply(strsplit(qcRawFiles,"/"),"[",2),"_f"),"[",1)
  
  rawReads[seq(2,length(rawReads),2)]
}

bosRawReads <- importRawReads(bosFolder)
#capraRawReads <- importRawReads(capraFolder)
equusRawReads <- importRawReads(equusFolder)
gallusRawReads <- importRawReads(gallusFolder)
ovisRawReads <- importRawReads(ovisFolder)
susRawReads <- importRawReads(susFolder)
```

## Number of raw reads

Visualize the raw reads

```{r boxplotR<awReads}
boxplot(bosRawReads, xlim=c(0.5,6.5), at=1)
#boxplot(capraRawReads, at=2, add=TRUE)
boxplot(equusRawReads, at=3, add=TRUE)
boxplot(gallusRawReads, at=4, add=TRUE)
boxplot(ovisRawReads, at=5, add=TRUE)
boxplot(susRawReads, at=6, add=TRUE)
axis(1, at=1:6, label=species)
```

## STAR mapping statistics

Import the STAR mapping statistics, Uniquely hits (black), multimapping (blue), too short (red)

```{r}
plotSTAR <- function(dir, main, boxplot=FALSE){
  starIn <- importSTARLog(file.path(dir, "BAM"))
  
  # Uniquely mapped reads
  uniqReads <- as.numeric(gsub("%","",as.vector(as.matrix(starIn$finalLog$uniqueReads[13,-1]))))
  mmReads <- as.numeric(gsub("%","",as.vector(as.matrix(starIn$finalLog$multiMapping[1,-1]))))
  tsReads <- as.numeric(gsub("%","",as.vector(as.matrix(starIn$finalLog$unmappedReads[3,-1]))))
  
  if(boxplot){
    boxplot(uniqReads, xlim=c(0.5,3.5), at=1, ylim=c(0,100), main=main)
    boxplot(mmReads, add=TRUE, at=2)
    boxplot(tsReads, add=TRUE, at=3)
    axis(1, at=c(1:3), labels=c("Uniq.", "Multim.", "Too short") ,las=2)
  } else {
    reads <- rbind(uniqReads, mmReads, tsReads)    
    barplot(reads, ylim=c(0,100), main=main, border=NA, space=0, col=c("black","blue","red"))
  }
}
```

```{r barplotsSTAR}
plotSTAR(bosFolder, "Bos taurus")
#plotSTAR(capraFolder, "Capra hircus")
plotSTAR(equusFolder, "Equus caballus")
plotSTAR(gallusFolder, "Gallus gallus")
plotSTAR(ovisFolder, "Ovis aries")
#plotSTAR(susFolder, "Sus scrofa")
```

```{r boxplotsSTAR}
plotSTAR(bosFolder, "Bos taurus", boxplot=TRUE)
#plotSTAR(capraFolder, "Capra hircus", boxplot=TRUE)
plotSTAR(equusFolder, "Equus caballus", boxplot=TRUE)
plotSTAR(gallusFolder, "Gallus gallus", boxplot=TRUE)
plotSTAR(ovisFolder, "Ovis aries", boxplot=TRUE)
#plotSTAR(susFolder, "Sus scrofa", boxplot=TRUE)
```

## Feature Counts statistics

Now some mapping and quantification statistics
```{r mappingStats}
# Import the featureCounts statistics
getFCSummary <- function(dir){
  fcFiles <- list.files(file.path(dir,"GTF", "Stringmerge_fc"), pattern="*.summary")
  fcSummary <- list()
  annotated <- c()
  unmapped <- c()
  unmapped.multi <- c()
  unmapped.ambi <- c()
  unannotated <- c()
  
  for(i in 1:length(fcFiles)){
    fcSummary[[i]] <- read.table(file.path(dir,"GTF", "Stringmerge_fc", fcFiles[i]), skip=1)
    annotated[i] <- fcSummary[[i]][1,2]
    unmapped[i] <- fcSummary[[i]][2,2]
    unmapped.multi <- fcSummary[[i]][7,2]
    unannotated[i] <- fcSummary[[i]][10,2]
    unmapped.ambi <- fcSummary[[i]][12,2]
  }
  
  out <- list(fcSummary, annotated, unmapped, unmapped.multi, unmapped.ambi, unannotated)
  out
}

bosMappingStats <- getFCSummary(bosFolder)
#capraMappingStats <- getFCSummary(capraFolder)
equusMappingStats <- getFCSummary(equusFolder)
#gallusMappingStats <- getFCSummary(gallusFolder)
#ovisMappingStats <- getFCSummary(ovisFolder)
susMappingStats <- getFCSummary(susFolder)
```

And now visualize the percentages of the different mapping scenarios

```{r}
par(mfrow=c(2,3))
createFCPlot <- function(x1, main, boxplots=TRUE){
  x2 <- sapply(sapply(x1[[1]],"[",2),sum)
  y1 <- x1[[2]]/x2
  y2 <- x1[[3]]/x2
  y3 <- x1[[4]]/x2
  y4 <- x1[[5]]/x2
  y5 <- x1[[6]]/x2
  if(boxplots){
    boxplot(y1, xlim=c(0.5,5.5), at=1, ylim=c(0,1), main=main)
    boxplot(y2, at=2, add=TRUE)
    boxplot(y3, at=3, add=TRUE)
    boxplot(y4, at=4, add=TRUE)
    boxplot(y5/x2, at=5, add=TRUE)
    axis(1, at=1:5, labels=c("Annot.", "Unmapped", "Multim.", "Ambi.", "Unannot."))  
  } else {
    stats <- rbind(y1,y2,y3,y4,y5)
    barplot(stats, ylim=c(0,5), main=main, border=NA, space=0,col=1:5)
  }
}
```

```{r plotFCplots}
createFCPlot(bosMappingStats, "Bos taurus")
#createFCPlot(capraMappingStats, "Capra hircus")
createFCPlot(equusMappingStats, "Equus caballus")
#createFCPlot(gallusMappingStats, "Gallus gallus")
#createFCPlot(ovisMappingStats, "Ovis aries")
createFCPlot(susMappingStats, "Sus scrofa")
```

# Expression analysis

## Expression values

We first import the expression values we got from featureCounts for the stringmerge output

```{r importExpressionValues}
getExpressions <- function(dir, lncRNA, annot=NULL, me){
  sm.expressions <- fread(file.path(dir, "GTF", "Stringmerge_fc.csv"))
  colnames(sm.expressions) <- c(colnames(sm.expressions)[1:6],gsub(".bam", "", sapply(strsplit(colnames(sm.expressions)[-c(1:6)],".*/"),"[",2)))
  RPK <- sm.expressions[,7:ncol(sm.expressions)]
  RPK <- RPK/(sm.expressions$Length/1000)
  PMscaling <- apply(RPK,2,sum)/1000000
  TPM <- t(t(RPK) /PMscaling)

  FEELnc <- TPM[is.element(sm.expressions$Geneid,lncRNA),]
  
  ENS.protein_coding <- NULL
  ENS.lncRNA <- NULL
  if(!is.null(annot)){
    annot.pc <- annot$gene_id[annot$gene_biotype=="protein_coding"]
    ENS.protein_coding <- TPM[is.element(sm.expressions$Geneid,annot.pc),]
    annot.lnc <- annot$gene_id[annot$gene_biotype=="lncRNA"]
    ENS.lncRNA <- TPM[is.element(sm.expressions$Geneid,annot.lnc),]
  }
  
  out <- list(counts = sm.expressions,
              TPM = TPM,
              RPK = RPK,
              FEELnc = FEELnc,
              ENS.protein_coding = ENS.protein_coding,
              ENS.lncRNA = ENS.lncRNA, 
              PMscaling = PMscaling)
  out
}
```

```{r createAnnots}
bosExpressions <- getExpressions(bosFolder, bosLNCIDs, bosAnnot, me=bosMSTR2Annot)
#capraExpressions <- getExpressions(capraFolder, capraLNCIDs)
equusExpressions <- getExpressions(equusFolder, equusLNCIDs)
#gallusExpressions <- getExpressions(gallusFolder, gallusLNCIDs)
#ovisExpressions <- getExpressions(ovisFolder, ovisLNCIDs)
susExpressions <- getExpressions(susFolder, susLNCIDs)
```

## Principle Component Analysis

```{r pca}
plotPCA <- function(x, data, main){
  if(data == "TPM"){
    dataToUse <- x$TPM
  } else if(data == "FEELnc"){
    dataToUse <- x$FEELnc
  } else {
    stop("No such data!!!")
  }
  d.pca <- prcomp(dataToUse, center=TRUE, scale=TRUE)
  
  pairs(d.pca$x[,1:5], main=paste("PCA", data , main))
}

plotPCA(bosExpressions, "FEELnc", "Bos taurus")
#plotPCA(capraExpressions, "FEELnc", "Capra hircus")
plotPCA(equusExpressions, "FEELnc", "Equus caballus")
#plotPCA(gallusExpressions, "FEELnc", "Gallus gallus")
#plotPCA(ovisExpressions, "FEELnc", "Ovis aries")
plotPCA(susExpressions, "FEELnc", "Sus scrofa")
```

## t-SNE

Now try t-SNE

```{r tsne}
plotTSNE <- function(x, data, main, colours){
    if(data == "TPM"){
    dataToUse <- x$TPM
  } else if(data == "FEELnc"){
    dataToUse <- x$FEELnc
  } else {
    stop("No such data!!!")
  }

  newOrder <- c()
  for(i in 1:ncol(dataToUse)){
    newPos <- which(colours$run_accession==colnames(dataToUse)[i])
    ifelse(length(newPos)>0, newOrder[i] <- newPos, newOrder[i] <- NA)
  }
  
  coloursOrdered <- colours[newOrder,]
  colourCodes <- as.numeric(as.factor(coloursOrdered$tissue.cell.line))

  coloursToUse <- colourCodes %% 7 + 1
  symbolToUse <- colourCodes %/% 7 + 1  
  
  forLegend <- data.frame(tissue=coloursOrdered$tissue.cell.line, 
                          col=as.numeric(coloursToUse), 
                          sym=as.numeric(symbolToUse))
  
  forLegend <- unique(forLegend)
  
  set.seed(9)  

  tsne_model_1 <-  Rtsne(as.matrix(t(dataToUse)), check_duplicates=FALSE, perplexity=10, pca=TRUE, theta=0, dims=2)

  ## getting the two dimension matrix
  d_tsne_1 <- as.data.frame(tsne_model_1$Y)  

  par(mar=c(5.1, 4.1, 4.1, 10), xpd=TRUE)

  plot(d_tsne_1, main=paste("t-SNE", data, main), col=coloursToUse, pch=symbolToUse)
  
  # Add legend to top right, outside plot region
  legend("topright", inset=c(-0.7,0), legend=forLegend$tissue, col=forLegend$col, pch=forLegend$sym, title="Tissue")

}
```

```{r plotTSNS}
plotTSNE(x=bosExpressions, "FEELnc", "Bos taurus", colours=samples[samples$species=="cattle",])
#plotTSNE(capraExpressions, "FEELnc", "Capra hircus", colours=samples[samples$species=="goat",])
plotTSNE(equusExpressions, "FEELnc", "Equus caballus", colours=samples[samples$species=="horse",])
#plotTSNE(gallusExpressions, "FEELnc", "Gallus gallus", colours=samples[samples$species=="chicken",])
#plotTSNE(ovisExpressions, "FEELnc", "Ovis aries", colours=samples[samples$species=="sheep",])
plotTSNE(susExpressions, "FEELnc", "Sus scrofa", colours=samples[samples$species=="pig",])
```